qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/ClusteringRule.java

```diff
@@ -606,16 +610,40 @@ public File getSnapshotsDirectory(final Broker broker) {
     return new File(dataDir, RAFT_PARTITION_PATH + "/snapshots");
   }
 
-  public void waitForValidSnapshotAtBroker(final Broker broker) {
-    waitForValidSnapshotAtBroker(broker, 1);
+  public DbSnapshotMetadata waitForSnapshotAtBroker(final Broker broker) {
+    return waitForNewSnapshotAtBroker(broker, null);
   }
 
-  void waitForValidSnapshotAtBroker(final Broker broker, final int snapshotCount) {
+  DbSnapshotMetadata waitForNewSnapshotAtBroker(
+      final Broker broker, final DbSnapshotMetadata previousSnapshot) {
+    final var referenceToResult = new AtomicReference<>(Optional.<DbSnapshotMetadata>empty());
     final File snapshotsDir = getSnapshotsDirectory(broker);
     waitUntil(
-        () ->
-            Optional.ofNullable(snapshotsDir.listFiles()).map(f -> f.length).orElse(0)
-                >= snapshotCount);
+        () -> {
+          final File[] files = snapshotsDir.listFiles();
+          if (files == null || files.length == 0) {
+            return false;
+          }
+          final Optional<DbSnapshotMetadata> latestSnapshot =
+              Arrays.stream(files)
+                  .map(File::toPath)
+                  .map(DbSnapshotMetadata::ofPath)
+                  .flatMap(Optional::stream)
+                  .max(Comparator.naturalOrder());
+          if (previousSnapshot != null
```

Comment:
I think this would work also if the latest is the same as the previous right? It then returns and I think it is a bit misleading that it says `waitForNew..` ?

Reply:
No that is not correct. Either `previousSnapshot` is null, in which case it is enough to just find a snapshot. Or `previousSnapshot` is not null, in which case `latestSnapshot.map(previousSnapshot::compareTo).orElse(0)` needs to be `-1`. The compareTo will return 0 if the `previousSnapshot` and the `latestSnapshot` are the same.
I've tried an early return (with exceptional cases first) approach to this method. So only in the last return `return true` it will have actually found something and stop waiting.

Reply:
```
latestSnapshot.map(previousSnapshot::compareTo).orElse(0)
```

Is translated to:

```
if (latestSnapshot != null)
{

  return latestsnapshot.compareTo(previousSnapshot); // return 1 when latest bigger, 0 when equal, -1 when smaller
}
return 0;
```

Or is it `previousSnapshot.compareTo(latestSnapshot)` ? :sweat_smile:

Reply:
It's actually: `latestSnapshot.map(latestSnapshot -> previousSnapshot.compareTo(latestSnapshot)).orElse(0)`

Reply:
So, 
if latest is empty, the result 0. 
If previous is before latest, the result is -1.
If previous is the same as latest, the result is 0.
If previous is after latest, the result is 1.

Reply:
Only the case where previous is before the latest (or in other words, the latest is actually the newest), then it stops waiting

Reply:
Ok sorry I get it now. As you can see it was for me not easy to understand :sweat_smile:  I always find this streaming api and compare a bit confusing :sweat_smile: But might be just me. If it is actually previousSnapshot.compareTo(latestSnapshot) then yes it makes sense otherwise it would not :sweat_smile:

Reply:
Yeah, streaming api can be confusing (just remember that it is a method reference, so in this case it will apply the provided method with the values in the Optional/Stream as argument). 

I have to add that I also wrote it a little confusing, but I wanted to maintain a return early approach. Perhaps I can have another look at it, to make it less confusing.

Reply:
@Zelldon Would this be more readable?
```
final File[] files = snapshotsDir.listFiles();
if (files == null || files.length == 0) {
  return false;
}
final Optional<DbSnapshotMetadata> latestSnapshot =
    Arrays.stream(files)
        .map(File::toPath)
        .map(DbSnapshotMetadata::ofPath)
        .flatMap(Optional::stream)
        .max(Comparator.naturalOrder());
if (latestSnapshot.isEmpty()) {
  return false;
}
if (previousSnapshot != null && previousSnapshot.compareTo(latestSnapshot.get()) > -1) {
  return false;
}
// latestSnapshot is newer than previousSnapshot (or previous is null)
referenceToResult.set(latestSnapshot);
return true;
```

Reply:
Yes I think it is easier from my point of view

Reply:
Ok, changed it.